### PR TITLE
Fix Authentication with Internal server when using SAML/SSO and fix "relgoin" button.

### LINKED ISSUE #58: Chat authentication doesn't work properly
Hi,

I have a Nextcloud 12.0.2 server with jsxc 3.2.1 app. Nextcloud has user_saml enabled using an environment variable and also a LDAP connection for user information retrieval (like email, full name).

When i login to Nextcloud and open the chat app i get "No connection! relogin". If i click relogin and enter my user/pass i get "Authentication failed with the chat server. Maybe the password is wrong?" even if user/pass are correct. But after this if i reload the page the chat app just logs me in. This works even if i provide wrong user/pass on the relogin.

Could this be fixed so that the chat app logs users in automatically?

### PR DESCRIPTION
**Pull request by LEDfan**
TODO:
 - [x] test with normal login procedure
 - [x] test with real XMPP server
 - [x] fix TODO's in code
 - [ ] fix roster push when logging in as a new user

fix #58

### CODE REVIEW

**Comment by LEDfan** [comment]
I looked into `fix roster push when logging in as a new user` but I didn't find a good/reliable way to do this with the SSO/SAML app. The app doens't emits the normal events send by the Nextcloud server. 

@sualko can you have a look if this makes PR sense? Thanks!

File: js/ojsxc.js [hunk 1]
```diff
@@ -205,10 +205,12 @@
                 password: password
             },
             success: function(d) {
-                if (d.result === 'success' && d.data && d.data.serverType !== 'internal' && d.data.xmpp.url !== '' && d.data.xmpp.url !== null) {
+                if (d.result === 'success' && d.data && d.data.serverType !== 'internal' && d.data.xmpp.url !== '' && d.data.xmpp.url !== null) {                       jsxc.storage.setItem('serverType', d.data.serverType);
```
Comment (sualko) [inline js/ojsxc.js]: Is here a newline missing?

Comment (sualko) [inline js/ojsxc.js]: Is here a newline missing?

File: js/ojsxc.js [hunk 2]
```diff
@@ -205,10 +205,12 @@
...
                     cb(d.data);
                 } else if (d.data && d.data.serverType === 'internal') {
```
Comment (sualko) [inline js/ojsxc.js]: I think it make sense to use a "constant" or "enum" for all required server types.

File: js/ojsxc.js [hunk 3]
```diff
@@ -392,4 +398,63 @@
...
+           startInternalBackend();
+        }
+    });
+
+    function startInternalBackend() {
+       jsxc.bid = OC.currentUser.toLowerCase() + '@' + window.location.host;
+
+        jsxc.storage.setItem('jid', jsxc.bid + '/internal');
+        jsxc.storage.setItem('sid', 'internal');
+        jsxc.storage.setItem('rid', '123456');
```
Comment (sualko) [inline js/ojsxc.js]: I guess this is not required (should be handled by `jsxc.start`), but I will check it.

File: js/ojsxc.js [hunk 4]
```diff
@@ -392,4 +398,63 @@
...
+
+        jsxc.options.set('xmpp', {
+            url: OC.generateUrl('apps/ojsxc/http-bind')
+        });
+
+        jsxc.start(jsxc.bid + '/internal', 'internal', '123456');
+        jsxc.gui.restore();
+        jsxc.gui.roster.toggle(jsxc.CONST.SHOWN);
+        $(document).trigger('attached.jsxc');
```
Comment (sualko) [inline js/ojsxc.js]: The above 3 lines should be handled by `jsxc.start`, but as above I will check it.

File: js/ojsxc.js [hunk 5]
```diff
@@ -392,4 +398,63 @@
...
+    $(document).on('stateChange.jsxc', function _handler(event, state) {
+        if (state === jsxc.CONST.STATE.SUSPEND) {
+            /**
+             * The first time we go into suspend mode we check if we are using the internal backend.
+             * If this is the case and the user explicitly press the "login_without_chat" button when logging
+             * into Nextcloud we know we are using another authentication mechanism (like SAML/SSO) and thus have
+             * to manually start the connection.
+             */
+            $(document).off('stateChange.jsxc', _handler);
+            if (jsxc.storage.getItem('serverType') === null) {
```
Comment (sualko) [inline js/ojsxc.js]: The `serverType` is never removed, so this only works once.

File: js/ojsxc.js [hunk 6]
```diff
@@ -392,4 +398,63 @@
...
+                $.ajax({
+                    url: OC.generateUrl('apps/ojsxc/settings/servertype'),
+                    success: function (data) {
+                        jsxc.storage.setItem('serverType', data.serverType);
+                        jsxc.gui.showLoginBox = function(){};
```
Comment (sualko) [inline js/ojsxc.js]: I think this should only be overwritten if the internal server type is used.

File: js/ojsxc.js (hunk 5 above)
Comment (LEDfan) [reply js/ojsxc.js]: This is intentional, if the serverType is know we can directly check if its interanl here: https://github.com/nextcloud/jsxc.nextcloud/pull/67/files#diff-4f938c34e352f51c3a310a503db3e0f7R451

**Comment by sualko** [comment]
I got the following error in my js console:

```
jsxc.js:10163 Uncaught TypeError: Cannot read property 'jingle' of null
    at HTMLDocument.init (jsxc.js:10163)
    at HTMLDocument.dispatch (jquery.min.js:3)
    at HTMLDocument.r.handle (jquery.min.js:3)
    at Object.trigger (jquery.min.js:3)
    at Object.a.event.trigger (jquery-migrate.min.js:2)
    at HTMLDocument.<anonymous> (jquery.min.js:3)
    at Function.each (jquery.min.js:2)
    at a.fn.init.each (jquery.min.js:2)
    at a.fn.init.trigger (jquery.min.js:3)
    at startInternalBackend (ojsxc.js:430)
```

I will try to fix it and merge it afterwards.

**Comment by sualko** [comment]
@LEDfan I tried to simplify various things and additionally found a bug in the core which prevents the ui restoration. This new version now also stops reconnecting, when the user forced a disconnect. Can you test my changes? Beware that you have to use the latest commit to jsxc.

**Comment by LEDfan** [comment]
@sualko Thanks, looks good. I just tested it with Keberos and internal server, internal server and external XMPP and everything works as expected.

**Pull request merged**